In [1]:
import torch
import numpy as np

# contiguous()

메모리에 0/1/2/3/4 순으로 저장되어 있으면 contiguous

메모리에 0/1/2/\n/3/4 순으로 저장되어 있으면 non contiguous

In [2]:
A = torch.tensor([[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11]])
A

tensor([[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]])

In [3]:
A.is_contiguous()

True

In [5]:
A_t = A.T
A_t

tensor([[ 0,  4,  8],
        [ 1,  5,  9],
        [ 2,  6, 10],
        [ 3,  7, 11]])

In [6]:
A_t.is_contiguous()

False

In [7]:
A_t_contigous = A_t.contiguous()

In [8]:
A_t_contigous.is_contiguous()

True

즉 메모리에 저장되어 있는 순서는 같은데, 보여지는 정보만 달라진 것이다.

이때 A_t.contiguous()를 통해 보여지는 것과 메모리에 할당된 순서를 업데이트 가능하다.

# argmax()

In [12]:
A = torch.randint(10, (3, 4, 2)) # 10까지 (3, 4, 2)차원인 텐서 생성
A

tensor([[[7, 4],
         [3, 0],
         [6, 0],
         [3, 6]],

        [[3, 2],
         [3, 7],
         [5, 9],
         [8, 2]],

        [[2, 4],
         [3, 7],
         [7, 8],
         [8, 3]]])

In [15]:
torch.argmax(A, dim=0)
# dim=0은 채널 방향. 즉 2차원 텐서 특정 지점에서 3개의 채널을 비교해서 제일 값이 큰 채널의 인덱스를 반환

tensor([[0, 0],
        [0, 1],
        [2, 1],
        [1, 0]])

In [14]:
torch.argmax(A, dim=0).shape

torch.Size([4, 2])

In [16]:
torch.argmax(A, dim=0, keepdim = True)

tensor([[[0, 0],
         [0, 1],
         [2, 1],
         [1, 0]]])

In [17]:
torch.argmax(A, dim=0, keepdim = True).shape
# 제일 값이 큰 채널이 있는 텐서빼고 다 날림

torch.Size([1, 4, 2])

In [20]:
print(torch.argmax(A, dim=1))
print(torch.argmax(A, dim=1).shape)
# 높이 기준

tensor([[0, 3],
        [3, 2],
        [3, 2]])
torch.Size([3, 2])


In [21]:
torch.argmax(A, dim=1, keepdim = True)

tensor([[[0, 3]],

        [[3, 2]],

        [[3, 2]]])

In [23]:
torch.argmax(A, dim=1, keepdim = True).shape

torch.Size([3, 1, 2])

In [24]:
torch.argmax(A, dim=2)
# 너비 기준

tensor([[0, 0, 0, 1],
        [0, 1, 1, 0],
        [1, 1, 1, 0]])

In [25]:
torch.argmax(A, dim=2).shape

torch.Size([3, 4])

In [26]:
torch.argmax(A, dim=2, keepdim=True)

tensor([[[0],
         [0],
         [0],
         [1]],

        [[0],
         [1],
         [1],
         [0]],

        [[1],
         [1],
         [1],
         [0]]])

In [27]:
torch.argmax(A, dim=2, keepdim=True).shape

torch.Size([3, 4, 1])

0번째 축은 화면 안으로 들어가는 방향, 1번째 축은 높이 방향, 2번째 축은 가로 방향

In [30]:
B = torch.rand(3, 2)
B

tensor([[0.9165, 0.6731],
        [0.9941, 0.3027],
        [0.2350, 0.8714]])

In [31]:
torch.argmax(B, dim=0)

tensor([1, 2])

즉 여기서는 높이 방향은 .9941, .8714가 제일 크고, 높이 위치는 각각 1, 2번째 인덱스이다.

In [32]:
torch.argmax(B, dim=0, keepdim=True)

tensor([[1, 2]])

In [33]:
torch.argmax(B, dim=1)

tensor([0, 0, 1])

In [34]:
torch.argmax(B, dim=1, keepdim=True)

tensor([[0],
        [0],
        [1]])

# numpy 배열과 토치의 텐서 변환

numpy 배열에서 토치의 텐서로

In [35]:
A = np.random.rand(3, 100, 100)
torch.from_numpy(A)

tensor([[[0.2661, 0.1321, 0.2578,  ..., 0.5965, 0.5601, 0.9763],
         [0.9791, 0.4239, 0.8937,  ..., 0.2862, 0.6900, 0.6624],
         [0.0432, 0.6333, 0.6014,  ..., 0.6868, 0.3571, 0.4619],
         ...,
         [0.8063, 0.2974, 0.7452,  ..., 0.0546, 0.9343, 0.8274],
         [0.4745, 0.8952, 0.5187,  ..., 0.5554, 0.8581, 0.3449],
         [0.1129, 0.7825, 0.2729,  ..., 0.4650, 0.3233, 0.5135]],

        [[0.5016, 0.8979, 0.8700,  ..., 0.7622, 0.7759, 0.3228],
         [0.1159, 0.6503, 0.6079,  ..., 0.3766, 0.7496, 0.1899],
         [0.9687, 0.1023, 0.4013,  ..., 0.3056, 0.5203, 0.0033],
         ...,
         [0.4212, 0.3171, 0.8454,  ..., 0.3388, 0.8447, 0.0338],
         [0.6091, 0.6382, 0.9271,  ..., 0.8156, 0.6660, 0.4343],
         [0.2916, 0.9309, 0.4379,  ..., 0.4610, 0.6087, 0.3420]],

        [[0.1530, 0.2102, 0.5695,  ..., 0.9908, 0.6878, 0.1926],
         [0.9012, 0.2175, 0.2803,  ..., 0.3420, 0.9963, 0.7903],
         [0.1905, 0.7665, 0.8879,  ..., 0.1864, 0.5605, 0.

토치의 텐서에서 넘파이 배열로의 변환은 cpu device 상에서 이루어져야 한다.

In [36]:
B = torch.tensor([1, 2, 3])
B_np = B.detach().cpu().numpy()
B_np

array([1, 2, 3], dtype=int64)

# torch.unsqueeze()와 torch.squeeze()

In [37]:
tensor = torch.rand(5,5)
print(tensor.shape)

torch.Size([5, 5])


In [38]:
tensor = torch.unsqueeze(tensor, 0)
tensor.shape

# 크기가 1인 새로운 차원 생성

torch.Size([1, 5, 5])

In [39]:
tensor = torch.rand(5,5)
tensor = torch.unsqueeze(tensor, 1)

# 크기가 1인 새로운 차원을 1번째 차원 텐서에 생성

print(tensor.shape)

torch.Size([5, 1, 5])


In [41]:
tensor = torch.rand(5, 5)
tensor = torch.unsqueeze(tensor, 2)
print(tensor.shape)

torch.Size([5, 5, 1])


unsqueeze가 크기가 1인 차원을 생성하는 것이라면, squeeze는 크기가 1인 차원을 제거 하는 것이다.

In [42]:
tensor = torch.rand(1, 5, 5)
print(tensor.shape)

torch.Size([1, 5, 5])


In [43]:
tensor = torch.squeeze(tensor, 0)
print(tensor.shape)

torch.Size([5, 5])


In [44]:
tensor = torch.rand(1, 5, 5)
tensor = torch.squeeze(tensor, 1)
print(tensor.shape)
# 차원 텐서의 해당 인덱스에서 크기가 1이 아니라면 제거 안됨

torch.Size([1, 5, 5])


# Variable(data)

In [46]:
from torch.autograd import Variable

v1 = Variable(torch.rand(3), requires_grad = True)
v1

tensor([0.9546, 0.9395, 0.9020], requires_grad=True)

In [47]:
with torch.no_grad():
    v2 = Variable(torch.rand(3))
print(v2)

tensor([0.4008, 0.5433, 0.4036])


즉 v1에 해당되는 텐서는 학습이 적용되고, v2에 해당되는 텐서들은 학습에 적용 안됨.

# F.interpolate()와 nn.Upsample()

2 메서드 전부 작은 데이터의 개수를 크게 할때 이용

In [48]:
import torch.nn as nn
import torch.nn.functional as F

In [49]:
input = torch.arange(0, 16, dtype = torch.float32).reshape(1, 1, 4, 4)
input

tensor([[[[ 0.,  1.,  2.,  3.],
          [ 4.,  5.,  6.,  7.],
          [ 8.,  9., 10., 11.],
          [12., 13., 14., 15.]]]])

In [50]:
F.interpolate(input, scale_factor=2, mode='nearest')

tensor([[[[ 0.,  0.,  1.,  1.,  2.,  2.,  3.,  3.],
          [ 0.,  0.,  1.,  1.,  2.,  2.,  3.,  3.],
          [ 4.,  4.,  5.,  5.,  6.,  6.,  7.,  7.],
          [ 4.,  4.,  5.,  5.,  6.,  6.,  7.,  7.],
          [ 8.,  8.,  9.,  9., 10., 10., 11., 11.],
          [ 8.,  8.,  9.,  9., 10., 10., 11., 11.],
          [12., 12., 13., 13., 14., 14., 15., 15.],
          [12., 12., 13., 13., 14., 14., 15., 15.]]]])

scale_factor는 interpolate를 할 목표사이즈 이다.
여기서 텐서의 구성요소 값들이 2개 반복된 것을 알수있다.

In [51]:
F.interpolate(input, scale_factor=0.8, mode='nearest')

C:\ProgramData\Anaconda3\lib\site-packages\torch\nn\functional.py:3679: UserWarning: The default behavior for interpolate/upsample with float scale_factor changed in 1.6.0 to align with other frameworks/libraries, and now uses scale_factor directly, instead of relying on the computed output size. If you wish to restore the old behavior, please set recompute_scale_factor=True. See the documentation of nn.Upsample for details. 
  warnings.warn(


tensor([[[[ 0.,  1.,  2.],
          [ 4.,  5.,  6.],
          [ 8.,  9., 10.]]]])

In [52]:
F.interpolate(input, scale_factor=2, mode='bilinear', align_corners=False)

tensor([[[[ 0.0000,  0.2500,  0.7500,  1.2500,  1.7500,  2.2500,  2.7500,
            3.0000],
          [ 1.0000,  1.2500,  1.7500,  2.2500,  2.7500,  3.2500,  3.7500,
            4.0000],
          [ 3.0000,  3.2500,  3.7500,  4.2500,  4.7500,  5.2500,  5.7500,
            6.0000],
          [ 5.0000,  5.2500,  5.7500,  6.2500,  6.7500,  7.2500,  7.7500,
            8.0000],
          [ 7.0000,  7.2500,  7.7500,  8.2500,  8.7500,  9.2500,  9.7500,
           10.0000],
          [ 9.0000,  9.2500,  9.7500, 10.2500, 10.7500, 11.2500, 11.7500,
           12.0000],
          [11.0000, 11.2500, 11.7500, 12.2500, 12.7500, 13.2500, 13.7500,
           14.0000],
          [12.0000, 12.2500, 12.7500, 13.2500, 13.7500, 14.2500, 14.7500,
           15.0000]]]])

In [53]:
F.interpolate(input, scale_factor=2, mode='bilinear', align_corners=True) 
# 어느정도 제일 부드럽게 upsampling 됨

tensor([[[[ 0.0000,  0.4286,  0.8571,  1.2857,  1.7143,  2.1429,  2.5714,
            3.0000],
          [ 1.7143,  2.1429,  2.5714,  3.0000,  3.4286,  3.8571,  4.2857,
            4.7143],
          [ 3.4286,  3.8571,  4.2857,  4.7143,  5.1429,  5.5714,  6.0000,
            6.4286],
          [ 5.1429,  5.5714,  6.0000,  6.4286,  6.8571,  7.2857,  7.7143,
            8.1429],
          [ 6.8571,  7.2857,  7.7143,  8.1429,  8.5714,  9.0000,  9.4286,
            9.8571],
          [ 8.5714,  9.0000,  9.4286,  9.8571, 10.2857, 10.7143, 11.1429,
           11.5714],
          [10.2857, 10.7143, 11.1429, 11.5714, 12.0000, 12.4286, 12.8571,
           13.2857],
          [12.0000, 12.4286, 12.8571, 13.2857, 13.7143, 14.1429, 14.5714,
           15.0000]]]])

In [54]:
F.interpolate(input, size=(5, 3), mode='bilinear', align_corners=False)

tensor([[[[ 0.1667,  1.5000,  2.8333],
          [ 2.9667,  4.3000,  5.6333],
          [ 6.1667,  7.5000,  8.8333],
          [ 9.3667, 10.7000, 12.0333],
          [12.1667, 13.5000, 14.8333]]]])

interpolate를 할때 size나 scale_factor중 둘중 하나만 선택하여야 한다.

size는 정확한 크기, scale_factor는 몇 배수 인지 정해주는 것이다.

In [55]:
input = torch.arange(0, 16, dtype=torch.float32).reshape(1, 1, 4, 4)

In [56]:
m = nn.Upsample(scale_factor=2, mode = 'nearest')
m(input)
# F.interpolate(input, scale_factor=2, mode='nearest') 와 같다

tensor([[[[ 0.,  0.,  1.,  1.,  2.,  2.,  3.,  3.],
          [ 0.,  0.,  1.,  1.,  2.,  2.,  3.,  3.],
          [ 4.,  4.,  5.,  5.,  6.,  6.,  7.,  7.],
          [ 4.,  4.,  5.,  5.,  6.,  6.,  7.,  7.],
          [ 8.,  8.,  9.,  9., 10., 10., 11., 11.],
          [ 8.,  8.,  9.,  9., 10., 10., 11., 11.],
          [12., 12., 13., 13., 14., 14., 15., 15.],
          [12., 12., 13., 13., 14., 14., 15., 15.]]]])

In [57]:
m = nn.Upsample(scale_factor=0.8, mode='nearest')
m(input)
# F.interpolate(input, scale_factor=0.8, mode='nearest')

tensor([[[[ 0.,  1.,  2.],
          [ 4.,  5.,  6.],
          [ 8.,  9., 10.]]]])

In [58]:
m = nn.Upsample(scale_factor=2, mode='bilinear', align_corners=False)
m(input)
# F.interpolate(input, scale_factor=2, mode='bilinear', align_corners=False)

tensor([[[[ 0.0000,  0.2500,  0.7500,  1.2500,  1.7500,  2.2500,  2.7500,
            3.0000],
          [ 1.0000,  1.2500,  1.7500,  2.2500,  2.7500,  3.2500,  3.7500,
            4.0000],
          [ 3.0000,  3.2500,  3.7500,  4.2500,  4.7500,  5.2500,  5.7500,
            6.0000],
          [ 5.0000,  5.2500,  5.7500,  6.2500,  6.7500,  7.2500,  7.7500,
            8.0000],
          [ 7.0000,  7.2500,  7.7500,  8.2500,  8.7500,  9.2500,  9.7500,
           10.0000],
          [ 9.0000,  9.2500,  9.7500, 10.2500, 10.7500, 11.2500, 11.7500,
           12.0000],
          [11.0000, 11.2500, 11.7500, 12.2500, 12.7500, 13.2500, 13.7500,
           14.0000],
          [12.0000, 12.2500, 12.7500, 13.2500, 13.7500, 14.2500, 14.7500,
           15.0000]]]])

In [59]:
m = nn.Upsample(size=(5, 3), mode='bilinear', align_corners=False)
m(input)
# F.interpolate(input, size=(5, 3), mode='bilinear', align_corners=False)

tensor([[[[ 0.1667,  1.5000,  2.8333],
          [ 2.9667,  4.3000,  5.6333],
          [ 6.1667,  7.5000,  8.8333],
          [ 9.3667, 10.7000, 12.0333],
          [12.1667, 13.5000, 14.8333]]]])

# 모델 block을 쌓기 - Module, Sequential, ModuleList, ModuleDict

Module : 여러 개의 작은 블록으로 구성된 큰 볼록이 있을 때

Sequential : 레이어에서 작은 블록을 만들고 싶을 때

ModuleList : 일부 레이어 또는 빌딩 블록을 반복하면서 어떤 작업을 해야 할때

ModuleDict : 모델의 일부 블록을 매개 변수화 해야하는 경우

## Module

In [60]:
class CNNClassifier(nn.Module):
    def __init__(self, in_c, n_classes):
        super().__init__()
        self.conv1 = nn.Conv2d(in_c, 32, kernel_size=3, stride=1, padding=1)
        self.bn1 = nn.BatchNorm2d(32)
        
        self.conv2 = nn.Conv2d(32, 64, kernel_size=3, stride=1, padding=1)
        self.bn2 = nn.BatchNorm2d(32)
        
        self.fc1 = nn.Linear(32*28*28, 1024)
        self.fc2 = nn.Linear(1024, 10)
        
    def forward(self, x):
        x = self.conv1(x)
        x = self.bn1(x)
        x = F.relu(x)
        
        x = self.conv2(x)
        x = self.bn2(x)
        x = F.relu(x)
        
        x = x.view(x.size(0), -1) # Flatten
        
        x = self.fc1(x)
        x = F.sigmoid(x)
        x = self.fc2(x)
        
        return x

## Sequential

In [61]:
class CNNClassifier(nn.Module):
    def __init__(self, in_c, n_classes):
        super().__init__()
        self.conv_block1 = nn.Sequential(
            nn.Conv2d(in_c, 32, kernel_size=3, stride=1, padding=1),
            nn.BatchNorm2d(32),
            nn.ReLU()
        )
        
        self.conv_block2 = nn.Sequential(
            nn.Conv2d(32, 64, kernel_size=3, stride=1, padding=1),
            nn.BatchNorm2d(64),
            nn.ReLU()
        )
        
        self.decoder = nn.Sequential(
            nn.Linear(32 * 28 * 28, 1024),
            nn.Sigmoid(),
            nn.Linear(1024, n_classes)
        )
    def forward(self, x):
        x = self.conv_block1(x)
        x = self.conv_block2(x)

        x = x.view(x.size(0), -1) # flat
        
        x = self.decoder(x)
        
        return x

보통 Sequential을 이용하면 다음과 같이 코드를 쓴다.

In [62]:
def conv_block(in_f, out_f, *args, **kwargs):
    return nn.Sequential(
        nn.Conv2d(in_f, out_f, *args, **kwargs),
        nn.BatchNorm2d(out_f),
        nn.ReLU()
    )

class CNNClassifier(nn.Module):
    def __init__(self, in_c, n_classes):
        super().__init__()
        self.encoder = nn.Sequential(
            conv_block(in_c, 32, kernel_size=3, padding=1),
            conv_block(32, 64, kernel_size=3, padding=1)
        )

        self.decoder = nn.Sequential(
            nn.Linear(32 * 28 * 28, 1024),
            nn.Sigmoid(),
            nn.Linear(1024, n_classes)
        )

        
    def forward(self, x):
        x = self.encoder(x)
        x = x.view(x.size(0), -1) # flat
        x = self.decoder(x)
        return x

훨씬 코드가 간결해졌다.

conv_block이 늘어나면 그렇게 보기는 좋지 않아 보인다.

따라서 다음과 같이 코드를 쓸 수 있다.

In [64]:
class CNNClassifier(nn.Module):
    def __init__(self, in_c, n_classes):
        super().__init__()
        self.enc_sizes = [in_c, 32, 64]
        conv_blocks = [conv_block(in_f, out_f, kernel_size=3, padding=1) 
                       for in_f, out_f in zip(self.enc_sizes, self.enc_sizes[1:])]
        self.encoder = nn.Sequential(*conv_blocks)
        
        self.decoder = nn.Sequential(
            nn.Linear(32 * 28 * 28, 1024),
            nn.Sigmoid(),
            nn.Linear(1024, n_classes)
        )
        
    def forward(self, x):
        x = self.encoder(x)
        x = x.view(x.size(0), -1) # flat
        x = self.decoder(x)
        
        return x

다음과 같이 반복문을 이용하여 코드를 쓸 수도 있따.

그러나 반복문을 이용할때 주의 할 것은 input 과 output의 채널 수이다.(CNN 코드)

따라서 channel의 수는 리스트를 이용하여 미리 저장하면 된다.

In [65]:
a = [1, 2, 3, 4, 5]
b = [10, *a]
print(b)

[10, 1, 2, 3, 4, 5]


In [66]:
def conv_block(in_f, out_f, *args, **kwargs):
    return nn.Sequential(
        nn.Conv2d(in_f, out_f, *args, **kwargs),
        nn.BatchNorm2d(out_f),
        nn.ReLU()
    )

def dec_block(in_f, out_f):
    return nn.Sequential(
        nn.Linear(in_f, out_f),
        nn.Sigmoid()
    )

class CNNClassifier(nn.Module):
    def __init__(self, in_c, enc_sizes, dec_sizes,  n_classes):
        super().__init__()
        self.enc_sizes = [in_c, *enc_sizes]
        self.dec_sizes = [32 * 28 * 28, *dec_sizes]

        conv_blokcs = [conv_block(in_f, out_f, kernel_size=3, padding=1) 
                       for in_f, out_f in zip(self.enc_sizes, self.enc_sizes[1:])]
        self.encoder = nn.Sequential(*conv_blokcs)
        
        dec_blocks = [dec_block(in_f, out_f) 
                       for in_f, out_f in zip(self.dec_sizes, self.dec_sizes[1:])]
        self.decoder = nn.Sequential(*dec_blocks)
        
        self.last = nn.Linear(self.dec_sizes[-1], n_classes)
        
    def forward(self, x):
        x = self.encoder(x)
        x = x.view(x.size(0), -1) # flat
        x = self.decoder(x)
        return x

즉 CNN연산의 경우 Convolution을 진행후 BatchNormalization 진행을 하는 연산이 반복이 되는데, 이러한 경우 Pytorch에서 Sequential 형태로 코드 블록을 작성하면 깔끔하게 작성이 가능하다.

## ModuleList

In [67]:
class MyModule(nn.Module):
    def __init__(self, sizes):
        super().__init__()
        self.layers = nn.ModuleList([nn.Linear(in_f, out_f) for in_f, out_f in zip(sizes, sizes[1:])])
        self.trace = []
        
    def forward(self,x):
        # ModuleList 에서는 각 Module을 하나 하나 접근할 수 있습니다.
        for layer in self.layers:
            x = layer(x)
            self.trace.append(x)
        return x

model = MyModule([1, 16, 32])
model(torch.rand((4,1)))
[print(trace.shape) for trace in model.trace]

torch.Size([4, 16])
torch.Size([4, 32])


[None, None]

Sequentai의 경우 Sequential로 묶은 단위에 대해서는 forward 연산이 발생하여 완전히 한 단위로 움직인다.

그러니 ModuleList는 리스트 형태로 각 Module을 접근하여 사용가능하다.

여기서보면 for 문을 이용하여 다양한 nn.Linear()층을 불러오는 것을 볼 수 있다.

## ModuleDict

In [68]:
def conv_block(in_f, out_f, activation='relu', *args, **kwargs):
    
    activations = nn.ModuleDict([
                ['lrelu', nn.LeakyReLU()],
                ['relu', nn.ReLU()]
    ])
    
    return nn.Sequential(
        nn.Conv2d(in_f, out_f, *args, **kwargs),
        nn.BatchNorm2d(out_f),
        activations[activation]
    )

print(conv_block(1, 32,'lrelu', kernel_size=3, padding=1))
print(conv_block(1, 32,'relu', kernel_size=3, padding=1))

Sequential(
  (0): Conv2d(1, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (1): BatchNorm2d(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (2): LeakyReLU(negative_slope=0.01)
)
Sequential(
  (0): Conv2d(1, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (1): BatchNorm2d(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (2): ReLU()
)


즉 여기서 Dictionary 처럼 다양한 Layer를 선택해서 호출해서 사용가능하다.